test/unit_tests/arkode/C_serial/ark_test_adapt.c

```diff
@@ -0,0 +1,153 @@
+  ARKodeEvolve(arkode_mem, tout, y, &tret, ARK_ONE_STEP);
+
+  long int local_err_fails;
+  ARKodeGetNumErrTestFails(arkode_mem, &local_err_fails);
+  if (local_err_fails != 0)
+  {
+    fprintf(stderr, "Expected 0 local error failures at step %i but is %li\n",
+            step, local_err_fails);
+  }
+
+  const N_Vector err = N_VClone(y);
+  ARKodeGetEstLocalErrors(arkode_mem, err);
+  const sunrealtype err_norm = N_VMaxNorm(err);
+  N_VDestroy(err);
+
+  if (err_norm != 0)
+  {
+    fprintf(stderr, "Expected local error at step %i to be 0 but is %g\n", step,
+            err_norm);
+    return 1;
+  }
+
+  sunrealtype h_actual;
+  ARKodeGetCurrentStep(arkode_mem, &h_actual);
+  if (SUNRCompare(h_expected, h_actual))
+  {
+    fprintf(stderr, "Expected h at step %i to be %g but is %g\n", step,
+            h_expected, h_actual);
+    return 1;
+  }
+
+  return 0;
+}
+
+/* Take several steps solving y'=0 and check the max growth was attained */
+static int check_steps(void* const arkode_mem, const N_Vector y,
+                       const sunrealtype h0, const sunrealtype first_growth,
+                       const sunrealtype growth)
+{
+  ARKodeSetInitStep(arkode_mem, h0);
+  ARKodeSetMaxFirstGrowth(arkode_mem, first_growth);
+  ARKodeSetMaxGrowth(arkode_mem, growth);
+
+  sunrealtype h_expect = first_growth * h0;
+  /* Take enough steps to allow the controller history to fill up */
+  const int num_steps = 5;
+  for (int step = 1; step <= num_steps; step++)
+  {
+    const int retval = check_step(arkode_mem, y, h_expect, step);
+    if (retval != 0) { return retval; }
+    h_expect *= growth;
+  }
+
+  return 0;
+}
+
+int main()
+{
+  SUNContext sunctx;
+  int retval = SUNContext_Create(SUN_COMM_NULL, &sunctx);
+  if (retval != 0)
+  {
+    fprintf(stderr, "SUNContext_Create returned %i\n", retval);
+    return 1;
+  }
+
+  retval = SUNContext_PushErrHandler(sunctx, err_fn, NULL);
+  if (retval != 0)
+  {
+    fprintf(stderr, "SUNContext_PushErrHandler returned %i\n", retval);
+    return 1;
+  }
+
+  const N_Vector y = N_VNew_Serial(1, sunctx);
+  N_VConst(SUN_RCONST(1.0), y);
+
+  /* Forward integration from 0 */
+  void* arkode_mem = ERKStepCreate(f, SUN_RCONST(0.0), y, sunctx);
+  retval = check_steps(arkode_mem, y, SUN_RCONST(1.0e-4), SUN_RCONST(1234.0),
+                       SUN_RCONST(3.0));
+  if (retval != 0) { return retval; }
+
+  /* Backward integration from positive time */
+  ERKStepReInit(arkode_mem, f, SUN_RCONST(999.0), y);
+  retval = check_steps(arkode_mem, y, SUN_RCONST(-1.0e-2), SUN_RCONST(1.6),
+                       SUN_RCONST(2.3));
+  if (retval != 0) { return retval; }
+
+  /* Forward integration from a negative time */
+  ERKStepReInit(arkode_mem, f, SUN_RCONST(-999.0), y);
+  retval = check_steps(arkode_mem, y, SUN_RCONST(20.0), SUN_RCONST(1.0e5),
+                       SUN_RCONST(1.1e3));
+  if (retval != 0) { return retval; }
+
+  /* TODO(SBR): test non-default controllers */
```

Comment:
Does this `TODO` need to be addressed?

Reply:
Yes, I'll try to add a couple more tests